onnxruntime/src/tensor/ort_tensor.rs

```diff
@@ -37,38 +38,103 @@ where
-        assert_ne!(tensor_ptr, std::ptr::null_mut());
+        match T::tensor_element_data_type() {
+            TensorElementDataType::Float
+            | TensorElementDataType::Uint8
+            | TensorElementDataType::Int8
+            | TensorElementDataType::Uint16
+            | TensorElementDataType::Int16
+            | TensorElementDataType::Int32
+            | TensorElementDataType::Int64
+            | TensorElementDataType::Double
+            | TensorElementDataType::Uint32
+            | TensorElementDataType::Uint64 => {
+                // primitive data is already suitably laid out in memory; provide it to
+                // onnxruntime as is
+                let tensor_values_ptr: *mut std::ffi::c_void =
+                    array.as_mut_ptr() as *mut std::ffi::c_void;
+                assert_ne!(tensor_values_ptr, std::ptr::null_mut());
+
+                unsafe {
+                    call_ort(|ort| {
+                        ort.CreateTensorWithDataAsOrtValue.unwrap()(
+                            memory_info.ptr,
+                            tensor_values_ptr,
+                            (array.len() * std::mem::size_of::<T>()) as u64,
+                            shape_ptr,
+                            shape_len,
+                            T::tensor_element_data_type().into(),
+                            tensor_ptr_ptr,
+                        )
+                    })
+                }
+                .map_err(OrtError::CreateTensorWithData)?;
+                assert_ne!(tensor_ptr, std::ptr::null_mut());
+
+                let mut is_tensor = 0;
+                let status = unsafe { g_ort().IsTensor.unwrap()(tensor_ptr, &mut is_tensor) };
+                status_to_result(status).map_err(OrtError::IsTensor)?;
+            }
+            TensorElementDataType::String => {
+                // create tensor without data -- data is filled in later
+                unsafe {
+                    call_ort(|ort| {
+                        ort.CreateTensorAsOrtValue.unwrap()(
+                            allocator_ptr,
+                            shape_ptr,
+                            shape_len,
+                            T::tensor_element_data_type().into(),
+                            tensor_ptr_ptr,
+                        )
+                    })
+                }
+                .map_err(OrtError::CreateTensor)?;
 
-        let mut is_tensor = 0;
-        let status = unsafe { g_ort().IsTensor.unwrap()(tensor_ptr, &mut is_tensor) };
-        status_to_result(status).map_err(OrtError::IsTensor)?;
-        assert_eq!(is_tensor, 1);
+                // create null-terminated copies of each string, as per `FillStringTensor` docs
+                let null_terminated_copies: Vec<ffi::CString> = array
+                    .iter()
+                    .map(|elt| {
+                        let slice = elt
+                            .try_utf8_bytes()
```

Comment:
Is the `TypeToTensorElementDataType::try_utf8_bytes()` really useful? If we send a `String`/`&str` to the function, we are guaranteed to be UTF-8 and can simply use [`String::as_bytes()`](https://doc.rust-lang.org/std/string/struct.String.html#method.as_bytes), no?

Reply:
How would we know when we have a string, though, if not via a method on `TypeToTensorElementDataType`? We could make the method just panic when you call it on the wrong type, but that makes usages of panic-y code harder to scan for than the noisy `.expect()`. 🤷

Reply:
I'm not sure I understand you, but I think I understand my previous misunderstanding...

So the method `try_utf_bytes()` is applied on the `elt` object which is of type `T`, _not_ `String` nor `&str`. This is why you need a trait to define the behaviour.

My bad!

Reply:
I was just saying that same thing in a more confusing way. ;)